modules/ensemble/lib/framework/device.dart

```diff
@@ -13,22 +14,95 @@ import 'package:ensemble_ts_interpreter/invokables/invokable.dart';
 import 'package:flutter/cupertino.dart';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
 import 'package:get_it/get_it.dart';
 import 'package:intl/intl.dart';
 
-/// get device information as well as requesting device permissions
-class Device
-    with
-        Invokable,
-        MediaQueryCapability,
-        LocationCapability,
-        DeviceInfoCapability {
+class Device with Invokable, MediaQueryCapability, LocationCapability, DeviceInfoCapability, WidgetsBindingObserver {
   static final Device _instance = Device._internal();
+  static bool _isInitialized = false;
+  bool _isHandlingChange = false;
+  MediaQueryData? _lastReportedData;
+
+  static final StreamController<MediaQueryData> _deviceUpdateController = 
+      StreamController<MediaQueryData>.broadcast();
+  static Stream<MediaQueryData> get onDeviceUpdate => _deviceUpdateController.stream;
+  
+  Device._internal() {
+    if (!_isInitialized) {
+      print('🔄 Initializing Device observer...');
+      WidgetsBinding.instance.addObserver(this);
+      // Set up initial data
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        _handleMediaQueryChange();  // Call this to set initial state
+      });
+      _isInitialized = true;
+    }
+  }
+
+  factory Device() => _instance;
+
+  @override
+  void didChangeMetrics() {
+    print('📏 Metrics changed - handling update...');
+    if (!_isHandlingChange) {
```

Comment:
what is this flag for, preventing multiple concurrent call? There is no async code here so that's not going to happen.